kernel/src/main/scala/cats/kernel/Order.scala

```diff
@@ -220,15 +218,73 @@ object Order extends OrderFunctions[Order] with OrderToOrderingConversion {
     }
 
   def fromOrdering[A](implicit ev: Ordering[A]): Order[A] =
-    new Order[A] {
-      def compare(x: A, y: A): Int = ev.compare(x, y)
-
-      override def toOrdering: Ordering[A] = ev
-    }
+    OrderFromOrdering[A](ev)
 
   def fromComparable[A <: Comparable[A]]: Order[A] =
     new Order[A] {
       override def compare(x: A, y: A): Int =
         x.compareTo(y)
     }
 }
+
+/** Wrapper class used to track when we create an `Order` instance from an
+  * `Ordering` instance. This allows us to know that given `Order` instance is
+  * consistent with the given `Ordering` instance. This becomes important when
+  * dealing with structures which rely on `Ordering`, e.g. `SortedSet`.
+  */
+final private[kernel] class OrderFromOrdering[A] private (val value: Ordering[A]) extends Order[A] {
+  override def compare(x: A, y: A): Int = value.compare(x, y)
+
+  override def toOrdering: Ordering[A] = value
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case that: OrderFromOrdering[A] @unchecked =>
+        value == that.value
+      case _ =>
+        false
+    }
+
+  override def hashCode: Int =
+    value.hashCode
+}
+
+private[kernel] object OrderFromOrdering {
+  def apply[A](ev: Ordering[A]): Order[A] =
+    ev match {
+      case ev: OrderingFromOrder[A] =>
+        ev.value
+      case _ =>
+        new OrderFromOrdering(ev)
+    }
+}
+
+/** Wrapper class used to track when we create an `Ordering` instance from an
+  * `Order` instance. This allows us to know that given `Order` instance is
+  * consistent with the given `Ordering` instance. This becomes important when
+  * dealing with structures which rely on `Ordering`, e.g. `SortedSet`.
+  */
+final private[kernel] class OrderingFromOrder[A] private (val value: Order[A]) extends Ordering[A] {
```

Comment:
same question about why not case class here?

Reply:
Could it be `AnyVal` as well?

Reply:
Oh, disregard please. I see that it does not make a lot of sense to make it `AnyVal` according to the way it is used.